### ISSUE TITLE
Creating a system image fails in Julia 1.10.0-beta2 while it works with 1.9

### ISSUE THREAD

**Issue by ufechner7** [JuliaLang/julia#50729]
On Ubuntu Linux version 20.04 with 16GB physical RAM creating a system image using package compiler fails badly when using Julia 1.10. The problem is that the system runs out of memory. With "fails badly" I mean that my laptop gets stuck, the OOM killer is not getting activated. Probably the memory usage rises too fast for the OOM killer.

This should not happen. Before starting the multi-threaded system image creation Julia should check the available free RAM (including the swap file) and if less than 6GB are free it should only start one thread. Alternatively there could be a command line option that determines the number of threads to use.

I use the following script to create a system image:
```julia
using Pkg

@info "Loading packages ..."
using ModelingToolkit, OrdinaryDiffEq, DataInterpolations, LaTeXStrings, Timers, 
      PackageCompiler, ControlSystemsBase, Roots, JLD2, YAML, DataFrames, MAT, EnumX,
      FFTW, Revise

PackageCompiler.create_sysimage(
    [:ModelingToolkit, :OrdinaryDiffEq, :DataInterpolations, :LaTeXStrings, :Timers,
     :ControlSystemsBase, :Roots, :JLD2, :YAML, :DataFrames, :MAT, :EnumX, :FFTW, :Revise];
    sysimage_path="kps-image_tmp.so",
    precompile_execution_file=joinpath("test", "test_for_precompile.jl")
)
```

**Comment by vchuravy** [comment]
Cc: @pchintalapudi

**Comment by pchintalapudi** [comment]
What's the `versioninfo()` for the system?

> Before starting the multi-threaded system image creation Julia should check the available free RAM (including the swap file) and if less than 6GB are free it should only start one thread.

This 6GB number is going to vary by size of system image and number of threads (2 threads might not exceed that for even larger images). There is always the escape hatch where you cut it down to one thread (set the `JULIA_IMAGE_THREADS=N` environment variable).

**Comment by ufechner7** [comment]
> What's the `versioninfo()` for the system?
> 
```
julia> versioninfo()
Julia Version 1.10.0-beta1
Commit 6616549950e (2023-07-25 17:43 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, skylake)
  Threads: 1 on 8 virtual cores
```
RAN usage Julia 1.9.2: 10GB
Julia 1.10.9-beta1: 24 GB

Steange: Julia 1.10 needs 24 GB on a 4 core CPU, but also on a 16 core CPU.

**Comment by pchintalapudi** [comment]
Does it work when set with one thread? (JULIA_IMAGE_THREADS=1)

**Comment by ufechner7** [comment]
> Does it work when set with one thread? (JULIA_IMAGE_THREADS=1)

No. With 1 thread 1.10 needs 17.4 GB, while 1.9 needs only 10.1GB.

So we have two issues:
- Julia 1.10.0-beta1 needs 1.7 times the RAM for building a sysimage
- it uses multiple threads by default

with 2, 4 and 16 threads it always needs 24GB RAM.

**Comment by pchintalapudi** [comment]
With or without multithreading?

**Comment by ufechner7** [comment]
> With or without multithreading? --heap-size-hint may be helpful here if the reduction is that steep.

Without multithreading the needed size is 12.7 GB, worse than with Julia 1.9.2 which needs 10.1GB, but good enough to be able to build the image with 16GB RAM. 

And no, --heap-size-hint has no effect, but that is another bug: https://github.com/JuliaLang/julia/issues/50658

**Comment by MariusDrulea** [comment]
Same behavior in Windows 10. Works with 1.9, but fails with 1.10.0-beta1.
```
using PackageCompiler

PackageCompiler.create_sysimage(
    [:GLMakie];
    sysimage_path="C:\\temp\\sys_image_sample.so",
    precompile_execution_file=joinpath(@__DIR__, "precompile.jl")
)
```

precompile.jl
```
using GLMakie
```

The error:
```
[ Info: PackageCompiler: Done
✔ [02m:53s] PackageCompiler: compiling incremental system image
D:/z_installed_programs/julia-depot/artifacts/fdff308295487f361ef6e8dc2d27f5abe8a6eee9/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: 
Error: export ordinal too large: 98037
collect2.exe: error: ld returned 1 exit status
ERROR: LoadError: failed process: Process(setenv(`'D:\z_installed_programs\julia-depot\artifacts\fdff308295487f361ef6e8dc2d27f5abe8a6eee9\mingw64\bin\g++.exe' -m64 -march=x86-64 -shared '-LD:\z_installed_programs\julia-depot\juliaup\julia-1.10.0-beta1+0.x64.w64.mingw32\bin' '-LD:\z_installed_programs\julia-depot\juliaup\julia-1.10.0-beta1+0.x64.w64.mingw32\bin' -o 'C:\temp\sys_image_sample.so' -Wl,--whole-archive 'C:\Users\someuser\AppData\Local\Temp\jl_X65etsvar4.o' -Wl,--no-whole-archive -ljulia -ljulia-internal -lopenlibm -Wl,--export-all-symbols`,["CLEARCASE_GROUPS=DE\\ISM_CC_USER_GW;DE\\ism_clearcase_all_lc;DE\\ism_cc_cssim_lc;DE\\ism_cc_mtc_lc;DE\\ism_cc_snsecs_test_lc;DE\\cceam3associates;DE\\Domain Users", "WINDIR=C:\\WINDOWS", "PATH=C:\\Program Files\\Git\\mingw64\\bin;C:\\Program Files\\Git\\usr\\local\\bin;C:\\Program Files\\Git\\usr\\bin;C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI2\\wbin;C:\\tools\\Anaconda3;C:\\tools\\Anaconda3\\Library\\mingw-w64\\bin;C:\\tools\\Anaconda3\\Library\\usr\\bin;C:\\tools\\Anaconda3\\Library\\bin;C:\\tools\\Anaconda3\\Scripts;C:\\WINDOWS\\system32;C:\\WINDOWS;C:\\WINDOWS\\System32\\Wbem;C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\missioncontrol\\;C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\webstart\\;C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\bin;C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\jre\\bin;C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0;C:\\WINDOWS\\System32\\OpenSSH;C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\missioncontrol;C:\\Oracle\\Client32\\bin;C:\\Program Files\\7-Zip;C:\\Program Files\\Microsoft SQL Server\\130\\Tools\\Binn;C:\\Program Files\\Microsoft SQL Server\\Client SDK\\ODBC\\170\\Tools\\Binn;C:\\Program Files\\Git\\cmd;C:\\Program Files\\Git\\;C:\\Program Files\\Calibre2;C:\\Program Files\\CMake\\bin;C:\\tools\\mingw64\\bin;C:\\Users\\someuser\\.dnx\\bin;C:\\Program Files\\Microsoft DNX\\Dnvm;C:\\Program Files\\Microsoft SQL Server\\120\\Tools\\Binn;C:\\Program Files (x86)\\dotnet;C:\\Users\\someuser\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Program Files\\Git\\usr\\bin\\vendor_perl;C:\\Program Files\\Git\\usr\\bin\\core_perl;C:\\Program Files\\dotnet\\;C:\\texlive\\2022\\bin\\win32;C:\\Program Files\\NVIDIA Corporation\\cuDNN\\v8.1.1\\bin;c:\\TCC\\Base\\StartBTC\\startBTC;c:\\TCC\\itc2;C:\\Program Files (x86)\\RedHat\\java-1.8.0-openjdk\\webstart\\;C:\\Program Files (x86)\\RedHat\\java-1.8.0-openjdk\\bin;C:\\Program Files (x86)\\RedHat\\java-1.8.0-openjdk\\jre\\bin;C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\;C:\\Program Files\\Graphviz\\bin;C:\\Program Files\\Microsoft VS Code\\bin;C:\\Users\\someuser\\AppData\\Local\\Microsoft\\WindowsApps;c:\\batch;c:\\tools;;D:\\z_installed_programs\\julia-depot\\artifacts\\fdff308295487f361ef6e8dc2d27f5abe8a6eee9\\mingw64\\bin", "USERDOMAIN_ROAMINGPROFILE=EMEA", "REDHAT_JAVA_HOME=C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\", "NEXTHINK=C:\\Program Files\\Nexthink\\Collector", "SNC_LIB_32=C:\\Program Files (x86)\\SAP\\FrontEnd\\SecureLogin\\lib\\sapcrypto.dll", "LOCALAPPDATA=C:\\Users\\someuser\\AppData\\Local", "MOZ_LEGACY_PROFILES=1", "HOMEPATH=\\Users\\someuser"  …  "FPS_BROWSER_APP_PROFILE_STRING=Internet Explorer", "SSF_LIBRARY_PATH_64=C:\\Program Files\\SAP\\FrontEnd\\SecureLogin\\lib\\sapcrypto.dll", "PROGRAMW6432=C:\\Program Files", "HOME=C:\\Users\\someuser", "HTTPS_PROXY=http://127.0.0.1:3128", "HOMEDRIVE=C:", "OPENBLAS_MAIN_FREE=1", "PROCESSOR_ARCHITECTURE=AMD64", "TEMP=C:\\Users\\someuser\\AppData\\Local\\Temp", "COLORTERM=truecolor"]), ProcessExited(1)) [1]

Stacktrace:
 [1] pipeline_error
   @ Base .\process.jl:565 [inlined]
 [2] run(::Cmd; wait::Bool)
   @ Base .\process.jl:480
 [3] run
   @ PackageCompiler .\process.jl:477 [inlined]
 [4] run_compiler(cmd::Cmd; cplusplus::Bool)
   @ PackageCompiler D:\z_installed_programs\julia-depot\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:194
 [5] run_compiler
   @ PackageCompiler D:\z_installed_programs\julia-depot\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:190 [inlined]
 [6] create_sysimg_from_object_file(object_files::Vector{…}, sysimage_path::String; version::Nothing, compat_level::String, soname::Nothing)
   @ PackageCompiler D:\z_installed_programs\julia-depot\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:634
 [7] create_sysimg_from_object_file
   @ PackageCompiler D:\z_installed_programs\julia-depot\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:620 [inlined]
 [8] create_sysimage(packages::Vector{…}; sysimage_path::String, project::String, precompile_execution_file::String, precompile_statements_file::Vector{…}, incremental::Bool, filter_stdlibs::Bool, cpu_target::String, script::Nothing, sysimage_build_args::Cmd, include_transitive_dependencies::Bool, base_sysimage::Nothing, julia_init_c_file::Nothing, version::Nothing, soname::Nothing, compat_level::String, extra_precompiles::String)
   @ PackageCompiler D:\z_installed_programs\julia-depot\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:600
 [9] top-level scope
   @ D:\z_installed_programs\julia-depot\dev\mypackage.jl_1.10\src\training\sysimage\how2sysimage.jl:3
Some type information was truncated. Use `show(err)` to see complete types.
in expression starting at D:\z_installed_programs\julia-depot\dev\mypackage.jl_1.10\src\training\sysimage\how2sysimage.jl:3
```

**Comment by pchintalapudi** [comment]
That is a different issue, happening during linking rather than out of memory.

```
Error: export ordinal too large: 98037
collect2.exe: error: ld returned 1 exit status
```

This is probably related to us using external hidden symbols to link across compilation units in multithreaded image generation.

**Comment by vtjnash** [comment]
Why are there many exports? There should only be a small handful

**Comment by pchintalapudi** [comment]
Maybe windows ld counts external hidden symbols as exports?

**Comment by pchintalapudi** [comment]
@MariusDrulea can you check if your issue still happens on #50791 with `JULIA_IMAGE_THREADS=1`?

**Comment by MariusDrulea** [comment]
> @MariusDrulea can you check if your issue still happens on #50791 with `JULIA_IMAGE_THREADS=1`?

I noticed the PR got merged into the main branch. Is there a way to use the nightly binaries in juliaup?

**Comment by MariusDrulea** [comment]
I wonder how to easily test PR builds. One way to deal with this is to rename the _julia-commit_ folder to  _julia-1.10.0-beta1+0.x64.w64.mingw32_. This way I can test any PR. While ok-ish, is there a more direct way to do it in julia-up?
![image](https://github.com/JuliaLang/julia/assets/27725572/cb6af431-8f20-49b9-bc9a-d017cb41cbff)

**Comment by MariusDrulea** [comment]
Here is the test for https://github.com/JuliaLang/julia/pull/50752. Still fails. Note I haven't set `JULIA_IMAGE_THREADS=1`.
```
PS D:\z_installed_programs\julia-depot\dev\mtdnn_base.jl_1.10> julia --project=.
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.11.0-DEV.231 (2023-08-07)
 _/ |\__'_|_|_|\__'_|  |  pc/internalize/fb0bc12ffb (fork: 2 commits, 5 days)
|__/  

julia> include("./src/training\\sysimage\\how2sysimage.jl")
Precompiling project...
  143 dependencies successfully precompiled in 184 seconds. 318 already precompiled.
  1 dependency had warnings during precompilation:
┌ ImageSegmentation [80713f31-8817-5129-9cf8-209ff8fb23e1]
│  [pid 32484] waiting for IO to finish:
│   TYPE[FD/PID]       @UV_HANDLE_T->DATA
│   timer              @00000243fa512ab0->00000243858109d0
└  
[ Info: PackageCompiler: Executing D:\z_installed_programs\julia-depot\dev\mypackage.jl_1.10\src\training\sysimage\precompile.jl => C:\Users\someuser\AppData\Local\Temp\jl_packagecompiler_UPI2zv\jl_4C23.tmp
[ Info: PackageCompiler: Done
⣠ [00m:54s] PackageCompiler: compiling incremental system imageExternalized 69110 / 110102(6.276907e-01) globals
✔ [02m:19s] PackageCompiler: compiling incremental system image
D:/z_installed_programs/julia-depot/artifacts/fdff308295487f361ef6e8dc2d27f5abe8a6eee9/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: Error: export ordinal too large: 69175
collect2.exe: error: ld returned 1 exit status
ERROR: LoadError: failed process: Process(setenv(`'D:\z_installed_programs\julia-depot\artifacts\fdff308295487f361ef6e8dc2d27f5abe8a6eee9\mingw64\bin\g++.exe' -m64 -march=x86-64 -shared '-LD:\z_installed_programs\julia-depot\juliaup\julia-1.10.0-beta1+0.x64.w64.mingw32\bin' '-LD:\z_installed_programs\julia-depot\juliaup\julia-1.10.0-beta1+0.x64.w64.mingw32\bin' -o 'C:\temp\sys_image_sample.so' -Wl,--whole-archive 'C:\Users\someuser\AppData\Local\Temp\jl_obuHURQLyp.o' -Wl,--no-whole-archive -ljulia -ljulia-internal -lopenlibm -Wl,--export-all-symbols`,["CLEARCASE_GROUPS=DE\\ISM_CC_USER_GW;DE\\ism_clearcase_all_lc;DE\\ism_cc_cssim_lc;DE\\ism_cc_mtc_lc;DE\\ism_cc_snsecs_test_lc;DE\\cceam3associates;DE\\Domain Users", "WINDIR=C:\\WINDOWS", "PATH=C:\\Program Files\\Git\\mingw64\\bin;C:\\Program Files\\Git\\usr\\local\\bin;C:\\Program Files\\Git\\usr\\bin;C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI2\\wbin;C:\\tools\\Anaconda3;C:\\tools\\Anaconda3\\Library\\mingw-w64\\bin;C:\\tools\\Anaconda3\\Library\\usr\\bin;C:\\tools\\Anaconda3\\Library\\bin;C:\\tools\\Anaconda3\\Scripts;C:\\WINDOWS\\system32;C:\\WINDOWS;C:\\WINDOWS\\System32\\Wbem;C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\missioncontrol\\;C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\webstart\\;C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\bin;C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\jre\\bin;C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0;C:\\WINDOWS\\System32\\OpenSSH;C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\missioncontrol;C:\\Oracle\\Client32\\bin;C:\\Program Files\\7-Zip;C:\\Program Files\\Microsoft SQL Server\\130\\Tools\\Binn;C:\\Program Files\\Microsoft SQL Server\\Client SDK\\ODBC\\170\\Tools\\Binn;C:\\Program Files\\Git\\cmd;C:\\Program Files\\Git\\;C:\\Program 
Files\\Calibre2;C:\\Program Files\\CMake\\bin;C:\\tools\\mingw64\\bin;C:\\Users\\someuser\\.dnx\\bin;C:\\Program Files\\Microsoft DNX\\Dnvm;C:\\Program Files\\Microsoft SQL Server\\120\\Tools\\Binn;C:\\Program Files (x86)\\dotnet;C:\\Users\\someuser\\AppData\\Local\\Microsoft\\WindowsApps;C:\\Program Files\\Git\\usr\\bin\\vendor_perl;C:\\Program Files\\Git\\usr\\bin\\core_perl;C:\\Program Files\\dotnet\\;C:\\texlive\\2022\\bin\\win32;C:\\Program Files\\NVIDIA Corporation\\cuDNN\\v8.1.1\\bin;c:\\TCC\\Base\\StartBTC\\startBTC;c:\\TCC\\itc2;C:\\Program Files (x86)\\RedHat\\java-1.8.0-openjdk\\webstart\\;C:\\Program Files (x86)\\RedHat\\java-1.8.0-openjdk\\bin;C:\\Program Files (x86)\\RedHat\\java-1.8.0-openjdk\\jre\\bin;C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\;C:\\Program Files\\Graphviz\\bin;C:\\Program Files\\Microsoft VS Code\\bin;C:\\Users\\someuser\\AppData\\Local\\Microsoft\\WindowsApps;c:\\batch;c:\\tools;;D:\\z_installed_programs\\julia-depot\\artifacts\\fdff308295487f361ef6e8dc2d27f5abe8a6eee9\\mingw64\\bin", "USERDOMAIN_ROAMINGPROFILE=EMEA", "REDHAT_JAVA_HOME=C:\\Program Files\\RedHat\\java-1.8.0-openjdk\\", "NEXTHINK=C:\\Program Files\\Nexthink\\Collector", "SNC_LIB_32=C:\\Program Files (x86)\\SAP\\FrontEnd\\SecureLogin\\lib\\sapcrypto.dll", "LOCALAPPDATA=C:\\Users\\someuser\\AppData\\Local", "MOZ_LEGACY_PROFILES=1", "HOMEPATH=\\Users\\someuser" 
 …  "FPS_BROWSER_APP_PROFILE_STRING=Internet Explorer", "SSF_LIBRARY_PATH_64=C:\\Program Files\\SAP\\FrontEnd\\SecureLogin\\lib\\sapcrypto.dll", "PROGRAMW6432=C:\\Program Files", "HOME=C:\\Users\\someuser", "HTTPS_PROXY=http://127.0.0.1:3128", "HOMEDRIVE=C:", "OPENBLAS_MAIN_FREE=1", "PROCESSOR_ARCHITECTURE=AMD64", "TEMP=C:\\Users\\someuser\\AppData\\Local\\Temp", "COLORTERM=truecolor"]), ProcessExited(1)) [1]

Stacktrace:
  [1] pipeline_error
    @ Base .\process.jl:565 [inlined]
  [2] run(::Cmd; wait::Bool)
    @ Base .\process.jl:480
  [3] run
    @ PackageCompiler .\process.jl:477 [inlined]
  [4] run_compiler(cmd::Cmd; cplusplus::Bool)
    @ PackageCompiler D:\z_installed_programs\julia-depot\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:194
  [5] run_compiler
    @ PackageCompiler D:\z_installed_programs\julia-depot\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:190 [inlined]
  [6] create_sysimg_from_object_file(object_files::Vector{String}, sysimage_path::String; version::Nothing, compat_level::String, soname::Nothing)
    @ PackageCompiler D:\z_installed_programs\julia-depot\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:634
  [7] create_sysimg_from_object_file
    @ PackageCompiler D:\z_installed_programs\julia-depot\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:620 [inlined]
  [8] create_sysimage(packages::Vector{…}; sysimage_path::String, project::String, precompile_execution_file::String, precompile_statements_file::Vector{…}, incremental::Bool, filter_stdlibs::Bool, cpu_target::String, script::Nothing, sysimage_build_args::Cmd, include_transitive_dependencies::Bool, base_sysimage::Nothing, julia_init_c_file::Nothing, version::Nothing, soname::Nothing, compat_level::String, extra_precompiles::String)
    @ PackageCompiler D:\z_installed_programs\julia-depot\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:600
  [9] top-level scope
    @ D:\z_installed_programs\julia-depot\dev\mypackage.jl_1.10\src\training\sysimage\how2sysimage.jl:3
 [10] include(fname::String)
    @ Base.MainInclude .\client.jl:489
 [11] top-level scope
    @ REPL[3]:1
in expression starting at D:\z_installed_programs\julia-depot\dev\mypackage.jl_1.10\src\training\sysimage\how2sysimage.jl:3
Some type information was truncated. Use `show(err)` to see complete types.
```

**Comment by MariusDrulea** [comment]
I set up the `JULIA_IMAGE_THREADS=1`. The issue still persist for https://github.com/JuliaLang/julia/pull/50752
```
PS D:\z_installed_programs\julia-depot\dev\mypackage.jl_1.10> $env:JULIA_NUM_THREADS = 1
PS D:\z_installed_programs\julia-depot\dev\mypackage.jl_1.10> $env:JULIA_NUM_THREADS    
1
PS D:\z_installed_programs\julia-depot\dev\mypackage.jl_1.10> julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.11.0-DEV.231 (2023-08-07)
|__/                   |

julia> versioninfo()
Julia Version 1.11.0-DEV.231
Commit fb0bc12ffb (2023-08-07 22:52 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 12 × Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, skylake)
  Threads: 1 on 12 virtual cores
Environment:
  JULIA_DEPOT_PATH = D:\z_installed_programs\julia-depot
  JULIA_NUM_THREADS = 1

julia> exit()

PS D:\z_installed_programs\julia-depot\dev\mypackage.jl_1.10> julia --project=. .\src\training\sysimage\how2sysimage.jl
[ Info: PackageCompiler: Executing D:\z_installed_programs\julia-depot\dev\mypackage.jl_1.10\src\training\sysimage\precompile.jl => C:\Users\ddr5clj\AppData\Local\Temp\jl_packagecompiler_6FcFNG\jl_3A0F.tmp
[ Info: PackageCompiler: Done
⡆ [00m:51s] PackageCompiler: compiling incremental system imageExternalized 69152 / 110102(6.280722e-01) globals
✔ [02m:15s] PackageCompiler: compiling incremental system image
D:/z_installed_programs/julia-depot/artifacts/fdff308295487f361ef6e8dc2d27f5abe8a6eee9/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: 
Error: export ordinal too large: 69217
collect2.exe: error: ld returned 1 exit status
```

**Comment by MariusDrulea** [comment]
@pchintalapudi I have also tested https://github.com/JuliaLang/julia/pull/50791 with `JULIA_NUM_THREADS=1`. Unfortunately the error is still there. As you can see, the number of exports nearly doubled. 
```
✔ [02m:23s] PackageCompiler: compiling incremental system image
D:/z_installed_programs/julia-depot/artifacts/fdff308295487f361ef6e8dc2d27f5abe8a6eee9/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: 
Error: export ordinal too large: 110170
```

**Comment by MariusDrulea** [comment]
Update:
https://github.com/JuliaLang/julia/pull/50752 indeed works with  `JULIA_IMAGE_THREADS=1`. If I don't set this env_variable, it fails.
1.10-beta1 does not work with  `JULIA_IMAGE_THREADS=1`

**Comment by pchintalapudi** [comment]
#50791 is the change that makes `JULIA_IMAGE_THREADS=1` work for your case, and it's marked for backport, so it'll get in before 1.10 is released (unsure if it'll make it into beta2).

**Comment by pchintalapudi** [comment]
Yes, but #50752 is based off #50791, which has the actual change for JULIA_IMAGE_THREADS=1 here. As noted, without JULIA_IMAGE_THREADS=1 #50752 isn't helping your case enough.

**Comment by MariusDrulea** [comment]
This https://github.com/JuliaLang/julia/pull/50874 works just fine on my Windows machine.

**Comment by MariusDrulea** [comment]
> I wonder how to easily test PR builds. One way to deal with this is to rename the _julia-commit_ folder to _julia-1.10.0-beta1+0.x64.w64.mingw32_. This way I can test any PR. While ok-ish, is there a more direct way to do it in juliaup?

To answer my own question, I just have to copy the binaries of some commit into the juliaup folder for example. 
Then `juliaup link 590a63b299 .\julia-590a63b299\bin\julia.exe` will create the `590a63b299` channel to our desired julia binary.
I use it with 'julia +590a63b299'. See more here: https://github.com/JuliaLang/juliaup

![image](https://github.com/JuliaLang/julia/assets/27725572/322af65c-974a-44aa-9e74-c4f323ffd8e9)

**Comment by ufechner7** [comment]
Well, this ha not been completed. Please, re-open. There have been two issues mixed up in this issue, one Linux issue and one Windows issue. The original issue is not solved. Out of memory...

**Comment by KristofferC** [comment]
I think the solution, for now, is for you to just set `JULIA_IMAGE_THREADS=1` to get back the old serial behavior which AFAIU works well to create the sysimage.

**Comment by ufechner7** [comment]
I think the most simple way to resolve this regression is either to use one thread as default, or to use the number of threads passed as -p parameter ...

What do you think?

It is not user friendly to assume that everybody has 32GB of RAM...

**Comment by ufechner7** [comment]
> Not sure what -p (the number of julia processes started for distributed computing) has to > do with anything of this
Well, my philosophy is better safe than sorry. So by default use one thread, use an existing command line option like -p to specify the number of threads to use or add a new command line option specially for this purpose.

Alternatively check the free memory before starting a second thread.

I don't think that tuning Julia for speed and accepting that crashes might happen that cannot even be avoided with a command line parameter is a good idea...

**Comment by ufechner7** [comment]
> That would make it worse in all cases where one does not run out of memory, which is most cases.

How do you come to the conclusion that this is "most cases"? I was running out of memory with 16GB RAM, many people have only 4 GB or 8 GB...

**Comment by vchuravy** [comment]
The number of threads we used is being calculated here: https://github.com/JuliaLang/julia/blob/1c536dd5f4625493230f05bf21076e214fbfe6eb/src/aotcompile.cpp#L1411

From the versioninfo you posted it seems you have 8 threads with 32 GB of RAM. 

Could you observe how much memory Julia uses for your case? Does it succeed with `JULIA_IMAGE_THREADS=2`? How much memory is being used at peak for `JULIA_IMAGE_THREADS=1`, `JULIA_IMAGE_THREADS=2`, `JULIA_IMAGE_THREADS=4`.

I am not sure what would be a satisfactory resolution. We use the same code for both small-ish workloads like pkgimgs and there this heuristic has worked well.
It's also worked well for medium workloads like the Julia system image. Now PackageCompiler.jl creates a very large workload, but it is also the workload that would benefit the most from parallelism.

**Comment by ufechner7** [comment]
> 32 GB of RAM

The versioninfo() I posted is from my private laptop which has 16GB physical RAM.

**Comment by vchuravy** [comment]
Ah, sorry I misread the numbers then. We could maybe apply a heuristic like `total_memory() / 4GB == num_threads`?
I am hesitant to do anything that looks at current memory usage since that is very transient.

**Comment by ufechner7** [comment]
That would not have helped in my case. Only `total_memory() / 16GB` = number of threads would have helped, and that is too conservative.

**Comment by gbaraldi** [comment]
Operating systems usually don't show free ram, if there is ram available they should and do use it for caching stuff. Macos is an example where free memory is very often in the hundred of megabytes but it's data that the OS will store in the disk if another process needs.

**Comment by ufechner7** [comment]
MWE for reproducing this issue on Linux (Ubuntu) with 16GB RAM, e.g. a virtual machine with 16GB RAM and 8 cores:

Project.toml:
```
[deps]
ControlSystems = "a6e380b2-a6ca-5380-bf3e-84a91bcd477e"
DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
ModelingToolkit = "961ee093-0014-501f-94e3-6117800e7a78"
PackageCompiler = "9b87118b-4619-50d2-8e1e-99f35a4d4d9d"
```
create_sys_image.jl:
```
using Pkg

@info "Loading packages ..."
using ModelingToolkit, ControlSystems, DataFrames, PackageCompiler

@info "Creating sysimage ..."

PackageCompiler.create_sysimage(
    [:ModelingToolkit, :ControlSystems, :DataFrames];
    sysimage_path="sys-image.so",
    precompile_execution_file="test_for_precompile.jl"
)
```
test_for_precompile.jl:
```
using ModelingToolkit, ControlSystems, DataFrames

df = DataFrame([[1, 2], [0, 0]], [:a, :b])
```
Executing the test:
- reboot
- launch terminal
- execute `sudo swapoff -a`
- execute `julia --project -e "include(\"create_sys_image.jl\")"`

This crashes with the OOM killer on Ubuntu 23.04 and Julia 1.10-beta1. It works fine with Julia 1.9.2.

**Comment by elextr** [comment]
@ufechner7 why do you turn swap off?  When there is no swap the OOM killer is likely to be more aggressive since an actual system OOM is much more severe if none can be freed by swapping.  If you only do it for the purposes of an MWE thats fine, but in real life having lots of swap might in fact allow you to complete (slowly).

> Operating systems usually don't show free ram, if there is ram available they should and do use it for caching stuff.

Also I wonder how much IO cache is used?  I would expect generating a sysimage to do quite a bit of IO.

But to my mind the question is why does the memory usage increase significantly with number of threads?  Its doing the same total work, making the sysimage, so unless the multithreaded version duplicates many structures(?) I would not have expected multiple threads to use a huge amount more memory.  Or is it because multiple threads are generating garbage faster than GC can recycle it, so total memory usage goes up, or is it something else?

If its GC related thats #50658 about the GC needing to be more aggressive when approaching a memory hint.

**Comment by ufechner7** [comment]
I turn swap off to have 16GB of RAM to have a reproducible test case...

**Comment by elextr** [comment]
It would be good to see the total allocations and GC % for one and two thread cases that complete.

**Comment by MariusDrulea** [comment]
@ufechner7 I run your MWE and the largest RAM usage I have seen is 12.4GB, for that specific julia process. See the picture. I do have a 32GB machine. However, I agree with you that sysimage creation in julia 1.10 shall work by default also on 8GB and 16GB computers.


![Screenshot from 2023-08-17 16-20-09](https://github.com/JuliaLang/julia/assets/27725572/e96a88b1-22c2-46d2-8323-366f8990a957)

**Comment by ufechner7** [comment]
> I run your MWE

Did you use Windows? Memory usage on Windows and Linux will be different...

**Comment by ufechner7** [comment]
I have to admit, this MWE is really at the edge: It works if swap is enabled:
![Screenshot from 2023-08-17 20-58-35](https://github.com/JuliaLang/julia/assets/4235229/9d796930-0657-441c-a8f2-8ee8d57bdb45)

With swap disabled it crashes when the memory usage reaches 15.3 GB...

**Comment by ufechner7** [comment]
Julia 1.10.0-beta2 has a slightly lower memory consumption, so the MWE posted above does not fail any longer... I am working on a new MWE.

**Comment by elextr** [comment]
Well if it works and 1.9 works can you please collect the allocation statistics and % GC time for both with single thread vs two threads.

**Comment by ufechner7** [comment]
How can I collect allocation statistics? Do I need to start Julia with a specific parameter?

**Comment by ufechner7** [comment]
Julia 1.10.
```
1.19 GB at the beginning
13.4 GB after 2 min
14.0 GB after 2:30
14.2 GB after 3:30

3:57 finished
julia> @time include("create_sys_image.jl")
[ Info: Loading packages ...
[ Info: Creating sysimage ...
[ Info: PackageCompiler: Executing /home/ufechner/repos/OOM/test_for_precompile.jl => /tmp/jl_packagecompiler_MAlnUO/jl_8LXjzR
[ Info: PackageCompiler: Done
✔ [03m:57s] PackageCompiler: compiling incremental system image
321.803640 seconds (11.55 M allocations: 757.911 MiB, 0.14% gc time, 1.25% compilation time: 14% of which was recompilation)
```

Julia 1.9.2
```
1.19 GB at the beginning
6.8 GB after 2:00
7.7 GB after 5:00
5:26 min finished

julia> @time include("create_sys_image.jl")
[ Info: Loading packages ...
[ Info: Creating sysimage ...
[ Info: PackageCompiler: Executing /home/ufechner/repos/OOM/test_for_precompile.jl => /tmp/jl_packagecompiler_gnIRDg/jl_dLHOAx
[ Info: PackageCompiler: Done
✔ [05m:26s] PackageCompiler: compiling incremental system image
376.732690 seconds (27.71 M allocations: 1.710 GiB, 0.21% gc time, 0.80% compilation time: 24% of which was recompilation)
```

What you can clearly see is that the reported allocations are pretty irrelevant. At least with 1.10. I assume that a separate Julia instance is launched and we do not see its allocations...

If we assume that the memory usage at the beginning was by the OS and not by Julia, then Julia 1.9.2 needed 6.5GB RAM and Julia 1.10-beta2 needed 13 GB.

**Comment by vtjnash** [comment]
> I turn swap off to have exactly 16GB of RAM to have a reproducible test case...

Swap is required, as otherwise it is not possible to use all 16GB of RAM, and a considerable amount of physical memory must be wasted to deal with the overcounting needed for all of the virtual memory that otherwise does not require physical memory due to de-duplication. For compilation machines, we expect the machines to have considerable resources, and we don't expect to reduce that requirement as we know users can access bigger machines if necessary to run the build.
